middleware/middleware.go

```diff
@@ -46,3 +50,27 @@ func newTokenMiddleware(verify authentication.TokenAuthentication, extractors ..
 		})
 	}
 }
+
+// AuthFuncGRPC returns a new grpc_auth.AuthFunc to use with the gazebo-web authentication library.
+//
+// The passed in context.Context will contain the gRPC metadata.MD object (for header-based authentication) and
+// the peer.Peer information that can contain transport-based credentials (e.g. `credentials.AuthInfo`).
+//
+//	auth := authentication.New[...]()
+//
+//	srv := grpc.NewServer(
+//		grpc.StreamInterceptor(grpc_auth.StreamServerInterceptor(AuthFuncGRPC(auth))),
+//		grpc.UnaryInterceptor(grpc_auth.UnaryServerInterceptor(AuthFuncGRPC(auth))),
+//	)
```

Comment:
Is there any way for us to apply interceptors at the method level, instead of at the server level? The motivation for this question is that we will need to support services that display data without requiring authentication (e.g. a landing page that showcases features with data before the user has logged in).

The answer might just be to write separate gRPC servers with different interceptors, but I want to know if there is a more granular way to go about this.

Reply:
That's a good question, thank you for raising it. Your gRPC Service can implement the following interface to skip certain methods from having authentication.

```go
// ServiceAuthFuncOverride allows a given gRPC service implementation to override the global `AuthFunc`.
//
// If a service implements the AuthFuncOverride method, it takes precedence over the `AuthFunc` method,
// and will be called instead of AuthFunc for all method invocations within that service.
type ServiceAuthFuncOverride interface {
	AuthFuncOverride(ctx context.Context, fullMethodName string) (context.Context, error)
}
```

Reply:
Hmm, this is virtually still applying at the service level. I suppose we would have to have something like `PublicService` and `AuthenticatedService` internally and then compose them, which is similar to writing separate services.

Thanks for the clarification.

Reply:
That's true, but we can use the `fullMethodName` argument to avoid applying auth to certain methods.